### PR TITLE
convert digest macros to function

### PR DESCRIPTION
**Pull request by pearsonca**
Per #212 - converts `_store`... macros to functions.

### CODE REVIEW

**Comment by pearsonca** [comment]
This can just be done, but I also think there's a way here to pull lines 678-685 into the `_store`...

File: src/digest.c [hunk 1]
```diff
@@ -103,32 +103,52 @@ SEXP is_little_endian(void) {
...
-} else for (int j = 0; j < output_length; j++) {               \
-    snprintf(output + j * 2, 3, "%02x", (WHAT)[j]);            \
-}                                                              \
-
-#define decl_tmp32(WHAT) uint32_t tmp = (WHAT);
-#define decl_tmp64(WHAT) uint64_t tmp = (WHAT);
+void _store_from_char_ptr(
+    const unsigned char *hash, unsigned char *output,
+    const size_t output_length, const int leaveRaw
+) {
```
Comment (eddelbuettel) [inline src/digest.c:109]: Could I plead with you to have indentation follow the rest of the file, please?  I.e.

```c++
void _store_from_char_ptr(const unsigned char *hash, unsigned char *output,
                          const size_t output_length, const int leaveRaw) {
     // ...

Comment (pearsonca) [reply src/digest.c:109]: absolutely - my bad, should have caught that was the style choice.

**Comment by eddelbuettel** [comment]
Overall this looks pretty good, and I like that you are incremental.  Maybe we can tag pulling 678-685 in for another round to make things compartmentalized?

I may turn the reverse depdendency machine on, which I haven't for the previous PR yet, just to see where we are.  Off for a morning run now, will look more later.  Thanks for continuing to kick the can down the road!

**Comment by pearsonca** [comment]
> There's also a bit of benchmarking to be done to see if the SHA512 version of conversion to string output is faster than the snprintf version. Either way, the SHA512 output can start using the shared function - question is whether the shared function should do that version or not.

Have poked at this a bit - confirm "classic" vs sha512 conversion to string is identical.

Hashing 10K random alphanumeric "words" (10 alphanumbers each), 100 times: md5 is mean 0.17834 (0.02043231 sd) with classic, 0.16378 (0.01895324 sd) with sha512 conversion, or ~8% reduction, but not a clear difference (due to uncertainty).

Same setup, but with xxhash32 (where the hashing bit should be faster): 0.15089 (0.01595315 sd)  with classic vs 0.15346 (0.01753347 sd) with sha512 - so ~2% increase, but again unclear difference.

I have a general preference for do-things-the-same-way. Vaguely inclined to believe the SHA512 is version is actually faster, so consolidate to that version?

File: src/digest.c (hunk 1 above)
Comment (eddelbuettel) [reply src/digest.c:109]: Thanks. It is the aspect of the 'now common with the kids these days' default choice I like the least as it "blends" function arguments with the function-local variableas as everything is indented the same way.  I find that less readable / imposing extra cognitive load.  Ah well.

**Comment by pearsonca** [comment]
Aside: there seems to be almost no difference between the algoritm choices in terms of effective time as used in R. I suspect hash time would be more apparent hashing big inputs, with better amortization of the shared-R bit, but my particular use case is many-small-inputs.

Might be a case for some S3 specializations?

**Comment by eddelbuettel** [comment]
It's still early morning so I still do quite understand what you are actually after.  In general I am also in favour of consistent approaches especially on our integration side, and extreme care about meddling with upstream code.

**Comment by eddelbuettel** [comment]
In which case I prefer more transparent / more commonly understood code.  `snprintf()` wins here if the performance is a tie.

(Benchmarking is a black hole, of course, as this will vary between compilers, standard library implementations, etc.)

File: src/digest.c [hunk 2]
```diff
@@ -103,32 +103,52 @@ SEXP is_little_endian(void) {
...
+    if (leaveRaw) {
+        memcpy(output, hash, output_length);
+    } else for (int j = 0; j < output_length; j++) {
+// one char is 2 hex digits => converting to (0-9A-F)-chars == writing 2 spots
```
Comment (eddelbuettel) [inline src/digest.c:113]: Very helpful comment, also appreciated!.  My editor would align the `//` with code, though :wink:

**Review by eddelbuettel — CHANGES_REQUESTED** [review]
I think this can go in once you align the function signature.

File: src/digest.c [hunk 3]
```diff
@@ -102,49 +102,51 @@ SEXP is_little_endian(void) {
     return Rf_ScalarLogical(BYTE_ORDER == LITTLE_ENDIAN);
 }
 
-/* for any algorithm that has output as unsigned char*, can use _store_from_char_ptr */
-void _store_from_char_ptr(
-    const unsigned char *hash, unsigned char *output,
-    const size_t output_length, const int leaveRaw
+#ifndef USESHA512
+#define USESHA512 0
+#endif
```
Comment (eddelbuettel) [inline src/digest.c:107]: Good idea!

**Comment by eddelbuettel** [comment]
Oh, we forgot a ChangeLog entry.  Oh well.  Next PR or just stick one here, I have a pending commit with two more ORCIDs anyway...

**Comment by pearsonca** [comment]
mmmm, something like: `src/digest.c: switch output handling from macro to function; consolidate stringification loop approaches

**Comment by eddelbuettel** [comment]
Done in 0ca2463

**Pull request merged**